torch/csrc/distributed/rpc/tensorpipe_agent.cpp

```diff
@@ -133,36 +184,40 @@ TensorPipeAgent::~TensorPipeAgent() {
 
 void TensorPipeAgent::startImpl() {
   VLOG(1) << "RPC agent for " << workerInfo_.name_ << " is starting";
-  auto uvContext = std::make_shared<tensorpipe::transport::uv::Context>();
 
-  std::string uvAddress = guessAddress(*uvContext);
-  VLOG(1) << "RPC agent for " << workerInfo_.name_ << " is using address "
-          << uvAddress;
+  std::vector<std::string> addresses;
+  int lowestPriority = -1;
+  std::string lowestPriorityTransport;
 
-  context_->registerTransport(1, "tcp", std::move(uvContext));
-#ifdef TP_ENABLE_SHM
-  context_->registerTransport(
-      0, "shm", std::make_shared<tensorpipe::transport::shm::Context>());
-#endif
-  context_->registerChannel(
-      1, "basic", std::make_shared<tensorpipe::channel::basic::Context>());
-#ifdef TP_ENABLE_CMA
-  context_->registerChannel(
-      0, "cma", std::make_shared<tensorpipe::channel::cma::Context>());
-#endif
+  for (auto& key : TensorPipeTransportRegistry()->Keys()) {
+    std::unique_ptr<TransportRegistration> reg =
+        TensorPipeTransportRegistry()->Create(key);
+    if (reg->priority > lowestPriority) {
+      lowestPriority = reg->priority;
+      lowestPriorityTransport = key;
+    }
+    addresses.push_back(c10::str(key, "://", reg->address));
+    context_->registerTransport(
+        reg->priority, std::move(key), std::move(reg->transport));
+  }
 
-  std::vector<std::string> addresses = {"tcp://" + uvAddress};
-#ifdef TP_ENABLE_SHM
-  addresses.push_back(createUniqueShmAddr());
-#endif
+  for (auto& key : TensorPipeChannelRegistry()->Keys()) {
+    std::unique_ptr<ChannelRegistration> reg =
+        TensorPipeChannelRegistry()->Create(key);
+    context_->registerChannel(
```

Comment:
Where there be any error if the Transport presents but channel is absent (or vice versa)?

Reply:
Yes, there should be at least one transport and one channel defined. If that is not the case, then an error would be raised somewhere, but I suspect this is not currently handled very gracefully, so probably it would just end in an uncaught exception.

Note however that a transport is necessary already just to open a pipe: the client must specify an address to connect to, and that address is specific to the transport (for example, the address could be `uv://127.0.0.1:1234` or `shm://some_unix_domain_socket_name`). So for the user to be able to use one such address it must have registered `uv` or `shm` already.

As for channels, it is possible that registering no channels would work if none of the messages sent over the pipes contain tensors (i.e., they only contain their payload, and thus only operate on scalars). It's not something that is supported/recommended. Also, the `basic` channel should be safe to always keep installed, because all it does is just re-use the transport to act as a channel, and is thus a sort-of fallback.